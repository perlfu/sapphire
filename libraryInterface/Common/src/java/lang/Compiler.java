/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang;

import org.jikesrvm.classloader.RVMClass;
import org.jikesrvm.classloader.RVMMethod;

/**
 * Jikes RVM implementation of {@link java.lang.Compiler}
 */
public final class Compiler {
  private Compiler() {
  }
  public static boolean compileClass(Class<?> klass) {
    RVMClass rvmKlass = java.lang.JikesRVMSupport.getTypeForClass(klass).asClass();
    if (rvmKlass.isResolved()) {
      RVMMethod[] methods = rvmKlass.getDeclaredMethods();
      for (RVMMethod method : methods) {
        if (!method.isAbstract() && !method.isNative()) {
          method.compile();
        }
      }
      return true;
    } else {
      return false;
    }
  }
  public static boolean compileClasses(String klasses) {
    try {
      return compileClass(Class.forName(klasses));
    } catch (ClassNotFoundException e) {
      return false;
    }
  }
  public static Object command(Object arg) {
    return null;
  }
  public static void enable() {
  }
  public static void disable() {
  }
}
