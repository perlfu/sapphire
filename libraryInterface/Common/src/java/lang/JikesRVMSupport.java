/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang;

import java.security.ProtectionDomain;
import java.lang.instrument.Instrumentation;

import org.jikesrvm.classloader.VM_Type;

import org.vmmagic.pragma.*;

import org.jikesrvm.VM;              // for VerifyAssertions and _assert()
import org.jikesrvm.scheduler.VM_Thread;

/**
 * Library support interface of Jikes RVM
 */
public class JikesRVMSupport {

  public static void initializeInstrumentation(Instrumentation instrumenter) {
    VMClassLoader.setInstrumenter(instrumenter);
  }

  public static Class<?>[] getAllLoadedClasses() {
    return VMClassLoader.getAllLoadedClasses();
  }

  public static Class<?>[] getInitiatedClasses(ClassLoader classLoader) {
    return VMClassLoader.getInitiatedClasses(classLoader);
  }

  public static Class<?> createClass(VM_Type type) {
    return Class.create(type);
  }

  public static Class<?> createClass(VM_Type type, ProtectionDomain pd) {
    Class<?> c = Class.create(type);
    setClassProtectionDomain(c, pd);
    return c;
  }

  public static VM_Type getTypeForClass(Class<?> c) {
    return c.type;
  }

  public static void setClassProtectionDomain(Class<?> c, ProtectionDomain pd) {
    c.pd = pd;
  }

  /***
   * String stuff
   * */

  @Uninterruptible
  public static char[] getBackingCharArray(String str) {
    return str.value;
  }

  @Uninterruptible
  public static int getStringLength(String str) {
    return str.count;
  }

  @Uninterruptible
  public static int getStringOffset(String str) {
    return str.offset;
  }

  /***
   * Thread stuff
   * */
  public static Thread createThread(VM_Thread vmdata, String myName) {
    if (VM.VerifyAssertions) VM._assert(VM.runningVM);
    Thread bootThread = new Thread(new VMThread(vmdata), myName,
        vmdata.getPriority(), vmdata.isDaemonThread());
    bootThread.group = ThreadGroup.root;
    return bootThread;
  }

  public static VM_Thread getThread(Thread thread) {
    if (thread == null) {
      return null;
    } else if(thread.vmThread == null) {
      return null;
    } else {
      return thread.vmThread.vmdata;
    }
  }

  public static void threadDied(Thread thread) {
    thread.die();
  }
  public static Throwable getStillBorn(Thread thread) {
    return thread.stillborn;
  }
  public static void setStillBorn(Thread thread, Throwable stillborn) {
    thread.stillborn = stillborn;
  }
  /***
   * Enum stuff
   */
  @Uninterruptible
  public static int getEnumOrdinal(Enum<?> e) {
    return e.ordinal;
  }
  @Uninterruptible
  public static String getEnumName(Enum<?> e) {
    return e.name;
  }
}
