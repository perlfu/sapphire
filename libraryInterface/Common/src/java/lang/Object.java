/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang;

import org.jikesrvm.objectmodel.VM_ObjectModel;
import org.jikesrvm.scheduler.VM_Thread;
import org.jikesrvm.runtime.VM_Runtime;
import org.vmmagic.pragma.Pure;

/**
 * Jikes RVM implementation of {@link java.lang.Object}
 *
 * By convention, order methods in the same order
 * as they appear in the method summary list of Sun's 1.4 Javadoc API.
 */
public class Object {

  @SuppressWarnings({"PMD.ProperCloneImplementation","PMD.CloneMethodMustImplementCloneable","CloneDoesntCallSuperClone"})
  protected Object clone() throws CloneNotSupportedException {
    return VM_Runtime.clone(this);
  }

  public boolean equals(Object o) {
    return this == o;
  }

  @SuppressWarnings({"PMD.EmptyFinalizer","FinalizeDoesntCallSuperFinalize"})
  protected void finalize() throws Throwable {
  }

  @Pure
  public final Class<?> getClass() {
    return VM_ObjectModel.getObjectType(this).getClassForType();
  }

  @Pure
  public int hashCode() {
    return VM_ObjectModel.getObjectHashCode(this);
  }

  public final void notify() throws IllegalMonitorStateException {
    VM_Thread.notify(this);
  }

  public final void notifyAll() throws IllegalMonitorStateException {
    VM_Thread.notifyAll(this);
  }

  @Pure
  public String toString() {
    return getClass().getName() + "@" + Integer.toHexString(hashCode());
  }

  public final void wait() throws InterruptedException,
                                   IllegalMonitorStateException {
    VM_Thread.wait(this);
  }

  public final void wait(long time) throws InterruptedException,
                                            IllegalMonitorStateException,
                                            IllegalArgumentException {
    wait(time, 0);
  }

  public final void wait(long time, int frac)  throws InterruptedException,
                                                      IllegalMonitorStateException,
                                                      IllegalArgumentException {
    if (time >= 0 && frac >= 0 && frac < 1000000) {
      if (time == 0 && frac > 0) {
        time = 1;
      } else if (frac >= 500000) {
        time += 1;
      }
      if (time == 0) {
        VM_Thread.wait(this);
      } else {
        VM_Thread.wait(this, time);
      }
    } else {
      throw new IllegalArgumentException();
    }
  }
}
