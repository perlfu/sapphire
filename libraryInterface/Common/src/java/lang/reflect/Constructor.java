/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang.reflect;

import org.jikesrvm.classloader.*;
import org.jikesrvm.runtime.VM_Reflection;
import org.jikesrvm.runtime.VM_Runtime;

/**
 * Implementation of java.lang.reflect.Constructor for JikesRVM.
 *
 * By convention, order methods in the same order
 * as they appear in the method summary list of Sun's 1.4 Javadoc API.
 */
public final class Constructor<T> extends AccessibleObject
  implements GenericDeclaration, Member {
  final VM_Method constructor;

  // Prevent this class from being instantiated.
  @SuppressWarnings("unused")
  private Constructor() {
    constructor = null;
  }

  // For use by JikesRVMSupport
  Constructor(VM_Method m) {
    constructor = m;
  }

  public boolean equals(Object other) {
    if (other instanceof Constructor) {
      return constructor == ((Constructor<?>)other).constructor;
    } else {
      return false;
    }
  }

  @SuppressWarnings("unchecked")  // Type system needs to be bent a bit here
  public Class<T> getDeclaringClass() {
    return (Class<T>)constructor.getDeclaringClass().getClassForType();
  }

  public Class<?>[] getExceptionTypes() {
    VM_TypeReference[] exceptionTypes = constructor.getExceptionTypes();
    if (exceptionTypes == null) {
      return new Class[0];
    } else {
      return JikesRVMSupport.typesToClasses(exceptionTypes);
    }
  }

  public int getModifiers() {
    return constructor.getModifiers();
  }

  public String getName() {
    return getDeclaringClass().getName();
  }

  public Class<?>[] getParameterTypes() {
    return JikesRVMSupport.typesToClasses(constructor.getParameterTypes());
  }

  public int hashCode() {
    return getName().hashCode();
  }

  public boolean isSynthetic() {
    return constructor.isSynthetic();
  }

  public Object newInstance(Object[] args) throws InstantiationException,
                                                  IllegalAccessException,
                                                  IllegalArgumentException,
                                                  InvocationTargetException {
    // Check accessibility
    if (!constructor.isPublic() && !isAccessible()) {
      VM_Class accessingClass = VM_Class.getClassFromStackFrame(1);
      JikesRVMSupport.checkAccess(constructor, accessingClass);
    }

    // validate number and types of arguments to constructor
    VM_TypeReference[] parameterTypes = constructor.getParameterTypes();
    if (args == null) {
      if (parameterTypes.length != 0) {
        throw new IllegalArgumentException("argument count mismatch");
      }
    } else {
      if (args.length != parameterTypes.length) {
        throw new IllegalArgumentException("argument count mismatch");
      }
      for (int i = 0; i < parameterTypes.length; i++) {
        args[i] = JikesRVMSupport.makeArgumentCompatible(parameterTypes[i].resolve(), args[i]);
      }
    }

    VM_Class cls = constructor.getDeclaringClass();
    if (cls.isAbstract()) {
      throw new InstantiationException("Abstract class");
    }

    // Ensure that the class is initialized
    if (!cls.isInitialized()) {
      try {
        VM_Runtime.initializeClassForDynamicLink(cls);
      } catch (Throwable e) {
        ExceptionInInitializerError ex = new ExceptionInInitializerError();
        ex.initCause(e);
        throw ex;
      }
    }

    // Allocate an uninitialized instance;
    Object obj = VM_Runtime.resolvedNewScalar(cls);

    // Run the constructor on the instance.
    try {
      VM_Reflection.invoke(constructor, obj, args);
    } catch (Throwable e) {
      throw new InvocationTargetException(e);
    }
    return obj;
  }

  public String toString() {
    StringBuilder sb = new StringBuilder(128);
    Modifier.toString(getModifiers(), sb).append(' ');
    sb.append(getDeclaringClass().getName()).append('(');
    Class<?>[] c = getParameterTypes();
    if (c.length > 0) {
        sb.append(JikesRVMHelpers.getUserName(c[0]));
        for (int i = 1; i < c.length; i++) {
          sb.append(',').append(JikesRVMHelpers.getUserName(c[i]));
        }
      }
    sb.append(')');
    c = getExceptionTypes();
    if (c.length > 0) {
        sb.append(" throws ").append(c[0].getName());
        for (int i = 1; i < c.length; i++) {
          sb.append(',').append(c[i].getName());
        }
      }
    return sb.toString();
  }

  // Generics support

  public TypeVariable<?>[] getTypeParameters() {
    VM_Atom sig = constructor.getSignature();
    if (sig == null) {
      return new TypeVariable[0];
    } else {
      return JikesRVMHelpers.getTypeParameters(this, sig);
    }
  }

  public Type[] getGenericExceptionTypes() {
    VM_Atom sig = constructor.getSignature();
    if (sig == null) {
      return getExceptionTypes();
    } else {
      return JikesRVMHelpers.getGenericExceptionTypes(this, sig);
    }
  }

  public Type[] getGenericParameterTypes() {
    VM_Atom sig = constructor.getSignature();
    if (sig == null) {
      return getParameterTypes();
    } else {
      return JikesRVMHelpers.getGenericParameterTypes(this, sig);
    }
  }

  public String toGenericString() {
    StringBuilder sb = new StringBuilder(128);
    Modifier.toString(getModifiers(), sb).append(' ');
    addTypeParameters(sb, getTypeParameters());
    sb.append(getDeclaringClass().getName()).append('(');
    Type[] types = getGenericParameterTypes();
    if (types.length > 0) {
        sb.append(types[0]);
        for (int i = 1; i < types.length; ++i) {
          sb.append(',').append(types[i]);
        }
      }
    sb.append(')');
    types = getGenericExceptionTypes();
    if (types.length > 0) {
        sb.append(" throws ").append(types[0]);
        for (int i = 1; i < types.length; i++) {
          sb.append(',').append(types[i]);
        }
      }
    return sb.toString();
  }

  static void addTypeParameters(StringBuilder sb, TypeVariable<?>[] typeArgs) {
    if (typeArgs.length == 0)
      return;
    sb.append('<');
    for (int i = 0; i < typeArgs.length; ++i) {
        if (i > 0) {
          sb.append(',');
        }
        sb.append(typeArgs[i]);
      }
    sb.append("> ");
  }
}
