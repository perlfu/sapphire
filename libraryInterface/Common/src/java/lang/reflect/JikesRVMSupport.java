/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang.reflect;

import org.jikesrvm.objectmodel.VM_ObjectModel;
import org.jikesrvm.runtime.VM_Runtime;
import org.jikesrvm.classloader.*;

/**
 * Library support interface of Jikes RVM
 */
public class JikesRVMSupport {

  /**
   * Convert from "vm" type system to "jdk" type system.
   */
  static Class<?>[] typesToClasses(VM_TypeReference[] types) {
    Class<?>[] classes = new Class[types.length];
    for (int i = 0; i < types.length; i++) {
      classes[i] = types[i].resolve().getClassForType();
    }
    return classes;
  }

  // Make possibly wrapped method argument compatible with expected type
  // throwing IllegalArgumentException if it cannot be.
  //
  @SuppressWarnings({"UnnecessaryBoxing","PMD.IntegerInstantiation"})
  static Object makeArgumentCompatible(VM_Type expectedType, Object arg) {
    if (expectedType.isPrimitiveType()) {
      if (arg instanceof java.lang.Void) {
        if (expectedType.isVoidType()) return arg;
      } else if (arg instanceof java.lang.Boolean) {
        if (expectedType.isBooleanType()) return arg;
      } else if (arg instanceof java.lang.Byte) {
        if (expectedType.isByteType()) return arg;
        if (expectedType.isShortType()) return new Short((Byte) arg);
        if (expectedType.isIntType()) return new Integer((Byte) arg);
        if (expectedType.isLongType()) return new Long((Byte) arg);
      } else if (arg instanceof java.lang.Short) {
        if (expectedType.isShortType()) return arg;
        if (expectedType.isIntType()) return new Integer((Short) arg);
        if (expectedType.isLongType()) return new Long((Short) arg);
      } else if (arg instanceof java.lang.Character) {
        if (expectedType.isCharType()) return arg;
        if (expectedType.isIntType()) return new Integer((Character) arg);
        if (expectedType.isLongType()) return new Long((Character) arg);
      } else if (arg instanceof java.lang.Integer) {
        if (expectedType.isIntType()) return arg;
        if (expectedType.isLongType()) return new Long((Integer) arg);
      } else if (arg instanceof java.lang.Long) {
        if (expectedType.isLongType()) return arg;
      } else if (arg instanceof java.lang.Float) {
        if (expectedType.isFloatType()) return arg;
        if (expectedType.isDoubleType()) return new Double((Float) arg);
      } else if (arg instanceof java.lang.Double) {
        if (expectedType.isDoubleType()) return arg;
      }
    } else {
      if (arg == null) return arg; // null is always ok
      VM_Type actualType = VM_ObjectModel.getObjectType(arg);
      if (expectedType == actualType ||
          expectedType == VM_Type.JavaLangObjectType ||
          VM_Runtime.isAssignableWith(expectedType, actualType)) {
        return arg;
      }
    }
    throw new IllegalArgumentException();
  }

  public static Field createField(VM_Field m) {
    return new Field(m);
  }

  public static Method createMethod(VM_Method m) {
    return new Method(m);
  }

  @SuppressWarnings("unchecked") // Can't type-check this without <T> type<T>, which breaks javac
  public static Constructor<?> createConstructor(VM_Method m) {
    return new Constructor(m);
  }

  public static VM_Field getFieldOf(Field f) {
    return f.field;
  }

  public static VM_Method getMethodOf(Method f) {
    return f.method;
  }

  public static VM_Method getMethodOf(Constructor<?> f) {
    return f.constructor;
  }


  /**
   * Check to see if a method declared by the accessingClass
   * should be allowed to access the argument VM_Member.
   * Assumption: member is not public.  This trivial case should
   * be approved by the caller without needing to call this method.
   */
  public static void checkAccess(VM_Member member, VM_Class accessingClass) throws IllegalAccessException {
    VM_Class declaringClass = member.getDeclaringClass();
    if (member.isPrivate()) {
      // access from the declaringClass is allowed
      if (accessingClass == declaringClass) return;
    } else if (member.isProtected()) {
      // access within the package is allowed.
      if (declaringClass.getClassLoader() == accessingClass.getClassLoader() && declaringClass.getPackageName().equals(accessingClass.getPackageName())) return;

      // access by subclasses is allowed.
      for (VM_Class cls = accessingClass; cls != null; cls = cls.getSuperClass()) {
        if (accessingClass == declaringClass) return;
      }
    } else {
      // default: access within package is allowed
      if (declaringClass.getClassLoader() == accessingClass.getClassLoader() && declaringClass.getPackageName().equals(accessingClass.getPackageName())) return;
    }

    throw new IllegalAccessException("Access to "+member+" is denied to "+accessingClass);
  }

}
