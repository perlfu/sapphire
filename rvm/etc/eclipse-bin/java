#! /usr/bin/env bash
# -*- coding: iso-8859-1 ; mode: shell-script ;-*-
# (C) Copyright © IBM Corp. 2004
#
# $Id$
#
## This is a pretend "java" executable that will run Jikes RVM with
## settings that make it work to run Eclipse.
#
# @author Steven Augart
# @date   19 December 2004

ME="$(basename $0)"

## Load in the environment.

if ! [[ ${RVM_BUILD-} ]]; then
    echo >&2 "$ME: You must set RVM_BUILD in order to run this program."
    exit 2
fi
if ! [[ ${RVM_ROOT-} ]]; then
    echo >&2 "$ME: You must set RVM_ROOT in order to run this program."
    exit 2
fi

## All of the following stuff to set up the Xalan processor is no longer
## unnecessary, since the "runrvm" script now always set up Xalan if
## your configuration file told Jikes RVM where to find Xalan.

# if ! . ${RVM_BUILD}/environment.target; then
#     echo >&2 "$ME: RVM_BUILD is set to \`$RVM_BUILD', but I can
#  not find a file name \`environment.host' there.  That file should be there
#  for any properly configured Jikes RVM build directory."
#     exit 2
# fi

# ## Get XALAN_J_DIR.  Make sure it is sane.
# if ! [[ ${XALAN_J_DIR} ]]; then
#     echo >&2 "$ME: Your Jikes RVM configuration doesn't have XALAN_J_DIR set.
#  You absolutely have to have the Xalan XML parser installed in order to
#  run Eclipse on top of Jikes RVM."
#     exit 2
# fi

# xalan_bin="${XALAN_J_DIR}/bin"
# for jar_file in bsf.jar xalan.jar xml-apis.jar xercesImpl.jar; do
#     f="${xalan_bin}/${jar_file}"
#     if ! [[ -f $f ]]; then
# 	echo >&2 "$ME: Your Jikes RVM configuration set XALAN_J_DIR to
# \`$XALAN_J_DIR', but I can not find a file that should be there:
# \`${f}'.  Since your Xalan installation appears broken, I am giving up."
# 	exit 2
#     fi
# done
# unset jar_file f

# # If you are using a different version of Xerces, then
# # the XML parser factory settings are as follows:
# #
# # javax.xml.parsers.DocumentBuilderFactory
# #     = org.apache.xerces.jaxp.DocumentBuilderFactoryImpl
# # javax.xml.parsers.SAXParserFactory=
# #     = org.apache.xerces.jaxp.SAXParserFactoryImpl
# # XERCES_PROPS="-Djavax.xml.parsers.DocumentBuilderFactory=org.apache.xerces.jaxp.DocumentBuilderFactoryImpl -Djavax.xml.parsers.SAXParserFactory=org.apache.xerces.jaxp.SAXParserFactoryImpl"
# ## We'll assume you're just running using the Xerces that came with XALAN, 
# ## though.

# XALAN_PROPS=-Djavax.xml.transform.TransformerFactory=org.apache.xalan.processor.TransformerFactoryImpl 
# XML="${xalan_bin}"
# XALAN_CLASSPATH=$XML/xalan.jar:$XML/xml-apis.jar:$XML/xercesImpl.jar:$XML/bsf.jar
# export RVM_BOOTCLASSPATH_APPEND="${XALAN_CLASSPATH}"

XALAN_PROPS=
XERCES_PROPS=



# These are useful when debugging problems:
# VERBOSE_OPTS="-verbose:class -X:base:verbose=true -X:vm:stackTraceAtExit=true -X:vm:verboseStackTrace=1"
VERBOSE_OPTS=""


# -Dosgi.locking=null: Present because GNU Classpath 0.10 does not implement
# some of the java.nio package's locking functions.  This means that
# we have to tell Eclipse 3.0 not to try to perform locking.
# (The osgi.locking system property is new with Eclipse 3.0)

## If you have lots of memory, you can hike up the 400M (400 megabytes) by
## even more, for possible greater efficiency.  But 400M seems to be enough to 
## run Eclipse. 
exec rvm -Xms400M $VERBOSE_OPTS -Dosgi.locking=none $XALAN_PROPS $XERCES_PROPS "$@"
