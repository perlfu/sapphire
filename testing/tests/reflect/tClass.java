/*
 * (C) Copyright IBM Corp. 2001
 */
//$Id$
import java.util.*;
import java.lang.reflect.*;

/**
 * @author unascribed
 */
public class tClass
   {
   public static void gc()
      {
      System.gc();
      }

   public static String hello(String say_this)
      {
      System.out.println("before gc arg=" + say_this);
      gc();
      System.out.println("after  gc arg=" + say_this);
      return "And I Say Hello Back";
      }

   public static int iello(String say_this, int return_this)
      {
      System.out.println("before gc arg=" + say_this);
      gc();
      System.out.println("after  gc arg=" + say_this);
      return return_this;
      }

   public static long lello(String say_this, long return_this)
      {
      System.out.println("before gc arg=" + say_this);
      gc();
      System.out.println("after  gc arg=" + say_this);
      return return_this;
      }

   public static int jello(int return_this, String say_this, int junk, int more_junk)
      {
      System.out.println("before gc arg=" + say_this);
      gc();
      System.out.println("after  gc arg=" + say_this);
      return return_this;
      }

   public String vello(String say_this)
      {
      return "And I Say Vello Back";      
      }

   public tClass(String s)
      {
      System.out.println("tClass constructor called with "+s);
      }

   private tClass()
     {
       System.out.println("tClass private constructor called!" );
     }

   public static void main(String args[]) throws Exception
      {
      // Class.forName
      //
      Class c = Class.forName("tClass");
      System.out.println(c);
      try
         {
         Class c_not_found = Class.forName("NotAClassSoThrowAnExceptionPlease");
         }
      catch (ClassNotFoundException e)
         {
         System.out.println("caught ClassNotFoundException");
         }
      
      // -----------------------------  Class.isArray()
      //
      if (c.isArray()) System.out.println(c+" is an array????");
      else             System.out.println(c+" is not an array...good");

      // ------------------------------  Class.getConstructors
      //
      Constructor ctors[] = c.getConstructors();
      Arrays.sort(ctors, new Comparator() {
              public int compare(Object x, Object y) {
                  return x.toString().compareTo( y.toString() );
              }
          });

      System.out.println(c+" has "+ctors.length+" visible constructors");
      for (int i = 0; i < ctors.length; ++i)
         System.out.println("   " + i + ": " + ctors[i]);

      Constructor declaredCtors[] = c.getDeclaredConstructors();
      Arrays.sort(declaredCtors, new Comparator() {
              public int compare(Object x, Object y) {
                  return x.toString().compareTo( y.toString() );
              }
          });

      System.out.println(c+" has "+declaredCtors.length+" declared constructors");
      for (int i = 0; i < declaredCtors.length; ++i)
         System.out.println("   " + i + ": " + declaredCtors[i]);

      // Class.getMethods  Method.getName
      //
      Method methods[] = c.getMethods();
      Method hello = null;
      Method iello = null;
      Method lello = null;
      Method jello = null;
      Method vello = null;

      Method declaredMethods[] = c.getDeclaredMethods();
      Arrays.sort(declaredMethods, new Comparator() {
              public int compare(Object x, Object y) {
                  return x.toString().compareTo( y.toString() );
              }
          });

      System.out.println(c + " has a total number of methods: "
                         + methods.length );
      for(int i = 0; i < methods.length; i++)
      { 
        // dont print the methods out, signitures are different in
        // java and RVM libraries for java/lang/Object methods.
        // System.out.println( methods[i] );
        if (methods[i].getName().equals("hello")) hello = methods[i];
        if (methods[i].getName().equals("iello")) iello = methods[i];
        if (methods[i].getName().equals("lello")) lello = methods[i];
        if (methods[i].getName().equals("jello")) jello = methods[i];
        if (methods[i].getName().equals("vello")) vello = methods[i];
      }

      System.out.println( " Number of declared methods: " + declaredMethods.length );

      for(int i = 0; i < declaredMethods.length; i++)
        System.out.println( declaredMethods[i] );

      // ------------------------------  invoke methods taking String, returning ref (String)
      //
      if (hello==null) 
         {
         System.out.println("tClass.hello not found!");
         System.exit(1);
         }
      else
         {
         System.out.println("================= READY TO CALL: "+hello);
         }

      // Method.invoke
      //
      int n_calls = 3;  // loop to see if we can crash gc
      while(n_calls-- > 0)
         {
         String hello_args[] = {"I Say Hello to You!"};
         String result = (String)hello.invoke(null, hello_args);
         System.out.println(result);
         }

      // ------------------------------  invoke methods taking String,int; returning int
      //
      if (iello==null) 
         {
         System.out.println("tClass.iello not found!");
         System.exit(1);
         }
      else
         {
         System.out.println("================= READY TO CALL: "+iello);
         }

      // Method.invoke
      //
      n_calls = 3;  // loop to see if we can crash gc
      while(n_calls-- > 0)
         {
         Object iello_args[] = {"I Say Iello to You!", new Integer(99)};
         Integer result = (Integer)iello.invoke(null, iello_args);
         System.out.println("Does this>"+result+"< look like 99?");
         }

      // ------------------------------  invoke methods taking String,long; returning long
      //
      if (lello==null) 
         {
         System.out.println("tClass.lello not found!");
         System.exit(1);
         }
      else
         {
         System.out.println("================= READY TO CALL: "+lello);
         }

      // Method.invoke
      //
      n_calls = 3;  // loop to see if we can crash gc
      while(n_calls-- > 0)
         {
         Object lello_args[] = {"I Say Lello to You!", new Long(99)};
         Long result = (Long)lello.invoke(null, lello_args);
         System.out.println("Does this>"+result+"< look like 99?");
         }


      // ------------------------------  invoke methods taking String,int; returning int
      //
      if (jello==null) 
         {
         System.out.println("tClass.jello not found!");
         System.exit(1);
         }
      else
         {
         System.out.println("================= READY TO CALL: "+jello);
         }

      // Method.invoke
      //
      n_calls = 3;  // loop to see if we can crash gc
      while(n_calls-- > 0)
         {
         Object jello_args[] = { new Integer(99), "I Say Jello to You!", new Integer(95), new Integer(94)};
         Integer result = (Integer)jello.invoke(null, jello_args);
         System.out.println("Does this>"+result+"< look like 99?");
         }

       // ------------------------------  newInstance
      tClass tc = new tClass("Hi!");
      String initargs[] = {"I'm dynamic!"};
      tClass tc_dyn = (tClass)ctors[0].newInstance(initargs);

      // ------------------------------  invoke methods taking String, returning String, but now virtual
      //
      if (vello==null) 
         {
         System.out.println("tClass.vello not found!");
         System.exit(1);
         }
      else
         {
         System.out.println("================= READY TO CALL: "+vello);
         }

      // Method.invoke
      //
      n_calls = 3;  // loop to see if we can crash gc
      while(n_calls-- > 0)
         {
         String vello_args[] = {"I Say Vello to You!"};
         String result = (String)vello.invoke(tc_dyn, vello_args);
         System.out.println(result);
         }
  

      }
   

   //------ 

   }
