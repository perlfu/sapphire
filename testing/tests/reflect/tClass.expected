class tClass
caught ClassNotFoundException
class tClass is not an array...good
class tClass has 1 visible constructors
   0: public tClass(java.lang.String)
class tClass has 2 declared constructors
   0: private tClass()
   1: public tClass(java.lang.String)
class tClass has a total number of methods: 16
 Number of declared methods: 7
public java.lang.String tClass.vello(java.lang.String)
public static int tClass.iello(java.lang.String,int)
public static int tClass.jello(int,java.lang.String,int,int)
public static java.lang.String tClass.hello(java.lang.String)
public static long tClass.lello(java.lang.String,long)
public static void tClass.gc()
public static void tClass.main(java.lang.String[]) throws java.lang.Exception
================= READY TO CALL: public static java.lang.String tClass.hello(java.lang.String)
before gc arg=I Say Hello to You!
after  gc arg=I Say Hello to You!
And I Say Hello Back
before gc arg=I Say Hello to You!
after  gc arg=I Say Hello to You!
And I Say Hello Back
before gc arg=I Say Hello to You!
after  gc arg=I Say Hello to You!
And I Say Hello Back
================= READY TO CALL: public static int tClass.iello(java.lang.String,int)
before gc arg=I Say Iello to You!
after  gc arg=I Say Iello to You!
Does this>99< look like 99?
before gc arg=I Say Iello to You!
after  gc arg=I Say Iello to You!
Does this>99< look like 99?
before gc arg=I Say Iello to You!
after  gc arg=I Say Iello to You!
Does this>99< look like 99?
================= READY TO CALL: public static long tClass.lello(java.lang.String,long)
before gc arg=I Say Lello to You!
after  gc arg=I Say Lello to You!
Does this>99< look like 99?
before gc arg=I Say Lello to You!
after  gc arg=I Say Lello to You!
Does this>99< look like 99?
before gc arg=I Say Lello to You!
after  gc arg=I Say Lello to You!
Does this>99< look like 99?
================= READY TO CALL: public static int tClass.jello(int,java.lang.String,int,int)
before gc arg=I Say Jello to You!
after  gc arg=I Say Jello to You!
Does this>99< look like 99?
before gc arg=I Say Jello to You!
after  gc arg=I Say Jello to You!
Does this>99< look like 99?
before gc arg=I Say Jello to You!
after  gc arg=I Say Jello to You!
Does this>99< look like 99?
tClass constructor called with Hi!
tClass constructor called with I'm dynamic!
================= READY TO CALL: public java.lang.String tClass.vello(java.lang.String)
And I Say Vello Back
And I Say Vello Back
And I Say Vello Back
